Add engine API to the black-box design (#12)
argoproj/gitops-engine: specs/design-black-box.md

## Design Details

The proposed design is based on the PoC that contains draft implementation and provides the base idea of a target design. PoC is available in
[argo-cd#fargo/engine](https://github.com/alexmt/argo-cd/tree/fargo/engine) and [flux#fargo](https://github.com/alexmt/flux/tree/fargo) 

The GitOps engine API consists of three main packages:
* The `utils` package. It consist of loosely coupled packages that implements K8S resource diffing, health assessment, etc

> * The `app` package that Application data structure definition and CRUD client interface.

Reviewer: What would be the (breaking) impact for Argo CD to change this into `repo` / `Repository`? I think we already discussed this multiple times, but when we detach this engine from Argo the 'application' context does not make much sense anymore, as the engine itself and Flux, unlike Argo, are focused around working with repositories.

Reviewer 2: +1

Author: I think you are right. After abstracting the manifest generation most of the Application spec is not used by the engine. Now it is much easier to come up with the Golang interface and remove the dependency on the Application.

Going through the app controller code and trying to find which methods should be exposed by the interface.  

I would like to name the interface `DeployableUnit`. This seems neutral to both Argo CD and Flux. Does the name sounds good to you? @hiddeco , @2opremio

Reviewer: Whenever I talk about the action Flux performs with regard to cluster operations I talk about syncing (and updating image tag references), so for me `Syncable<Something>` would make more sense to me.

Reviewer 3: My vote was `ManagedRepository` :)

Author: I guess Application is very Argo specific and Repo is very Flux specific :) 

How about `SyncUnit`? Both Argo CD and Flux uses term `syncing` so `SyncUnit` would be understandable to users from both communities.

Author: Worked on draft changes in the engine and definitely looks better if we remove reference to the application. Updated design doc and removed Application from it.

I've tried using different names that include `Source` ( `SyncSource`, `TruthSource`  ) but It looks strange that Source has Source and Destination. 
e.g.

```
mySource := SyncSource{}
mySource.Source = ...
mySource.Destination = ...
```

So temporary using `SyncUnit` and totally open to change it

Author: Replace `SyncUnit` with `Sync<Term>`. @hiddeco will setup a poll in slack to decide the best name :)

* The `engine` package that leverages `utils` and uses provided set of Applications as a configuration source.

```
gitops-engine
|-- pkg
|   |-- utils
|   |   |-- diff   # provides Kubernetes resource diffing functionality
|   |   |-- kube   # provides utility methods to interact with Kubernetes

> |   |   |-- lue    # provides utility methods to run Lua scripts

Reviewer 2: ```suggestion
|   |   |-- lua    # provides utility methods to run Lua scripts
```

Reviewer 2: I think we went over this before (sorry), but why do we need Lua scripting in the core?

Author: Argo CD uses lua to customize health assessment.

Flux won't use it right now. It would be easy to add one more callback and remove the dependency on lua. Let me know if you want to do it please. I don't have strong opinion about it.

|   |   |-- apps   # provides utility methods to manipulate Applications (e.g. start sync operation, wait for sync operation, force reconciliation)
|   |   `-- health # provides Kubernetes resources health assessment
|   |-- app
|   |   |-- apis       # contains data structures that describe Application
|   |   `-- client     # contains client that provides Application CRUD operations
|   `-- engine     # the engine implementation
```

> The engine is responsible for reconciliation only which includes:

Reviewer 2: ```suggestion
The engine is responsible for the reconciliation of Kubernetes resources, which includes:
```

- Interacting with the Kubernetes API: load and maintain the cluster state; pushing required changes to the Kubernetes API.

> - Reconciliation logic: based on provides target resource definition find which resources should be updated/deleted/created.

Reviewer 2: I cannot parse this :S . typo?

Author: Looks like it was too early in the morning :) fixed.

- Syncing logic: determine the order in which resources should be modified; features like sync hooks, waves, etc.

The manifests generation is out of scope and should be implemented by the Engine consumer.

### Engine API

The engine API includes Application data structure/client and `Engine` golang interface that allows configuring reconciliation process:

> **Engine interface** - provides set of that allows updating reconciliation settings and subscribe to engine events.

Reviewer 2: set of .. functionality? features

Author: fixed

```golang
type Engine interface {
	// Run starts reconciliation loop using specified number of processors for reconciliation and operation execution.
	Run(ctx context.Context, statusProcessors int, operationProcessors int)

	// SetReconciliationSettings updates reconciliation settings
	SetReconciliationSettings(settings ReconciliationSettings)

	// OnBeforeSync registers callback that is executed before each sync operation.
	OnBeforeSync(callback func(appName string, tasks []SyncTaskInfo) ([]SyncTaskInfo, error)) Unsubscribe

	// OnSyncCompleted registers callback that is executed after each sync operation.
	OnSyncCompleted(callback func(appName string, state appv1.OperationState) error) Unsubscribe

> 	// OnClusterCacheInitialized registers that that is executed when cluster cache initialization is completed.

Reviewer 2: ```suggestion
	// OnClusterCacheInitialized registers callback that is executed when cluster cache initialization is completed.
```

> 	OnClusterCacheInitialized(callback func(server string)) Unsubscribe

Reviewer 2: Why is this needed?

Reviewer: Judging by the changes in the `fargo` branch of Flux, it seems to be used to kickstart the engine. https://github.com/alexmt/flux/commit/b4cc9b6a99966f75a03687a1b0599b1e8b5ef047#diff-6b48cc88be8e3f8560bbe695aa8b202fR90

Author: @hiddeco  is right. The callback is required to block `RunEngine` until cache is initialized and to make sure that Flux don't use the incomplete list of workloads.

Author: done. Added comment.

> 	// OnResourceUpdated registers that that is executed when cluster resource got updated.

Reviewer 2: ```suggestion
	// OnResourceUpdated registers callback that is executed when cluster resource got updated.
```

	OnResourceUpdated(callback func(cluster string, un *unstructured.Unstructured)) Unsubscribe

> 	// OnResourceRemoved registers that that is executed when cluster resource got removed.

Reviewer 2: ```suggestion
	// OnResourceRemoved registers a callback that is executed when cluster resource got removed.
```

	OnResourceRemoved(callback func(cluster string, key kube.ResourceKey)) Unsubscribe

	// OnAppEvent registers callback that is executed on every application event.
	OnAppEvent(callback func(app *appv1.Application, info EventInfo, message string)) Unsubscribe
}

type Unsubscribe func()
```

**ReconciliationSettings data structure** - holds reconciliation settings
```golang
type ReconciliationSettings struct {
	// AppInstanceLabelKey holds label key which is automatically added to every resource managed by the application
	AppInstanceLabelKey string
	// ResourcesFilter holds settigns which allows to configure list of managed resource APIs
	ResourcesFilter resource.ResourcesFilter

> 	// ResourceOverrides holds settings which customize resource diffing logic

Reviewer 2: How does this work? what should be put in the keys?

Author: Key is `group/kind`. Here is example of real life configuration:

```
    certmanager.k8s.io/Certificate:
      # Lua script for customizing the health status assessment
      health.lua: |
        hs = {}
        if obj.status ~= nil then
          # some logic here
        end
        hs.status = "Progressing"
        hs.message = "Waiting for certificate"
        return hs
```

Do you think it makes sense to replace string with `scheme.GroupKind` ?

Author: done

	ResourceOverrides map[string]appv1.ResourceOverride
}
```

**CredentialsStore** provides access to cluster credentials
```golang
type CredentialsStore interface {
	GetCluster(ctx context.Context, name string) (*appv1.Cluster, error)
	WatchClusters(ctx context.Context, callback func(event *ClusterEvent)) error
}
```

> **ManifestGenerator** an golang interface that must be implemented by the GitOps engine consumer.

Reviewer 2: ```suggestion
**ManifestGenerator** a Golang interface that must be implemented by the GitOps engine consumer.
```

```golang

> type ManifestResponse struct {

Reviewer: Something is off with the alignment here, I think it is tabs vs spaces.

Author: fixed

    // Generated manifests
    Manifests  []string
    // Resolved Git revision

> 	Revision   string

Reviewer 2: ```suggestion
    Revision   string
```
